primitives/runtime-interface/src/lib.rs

```diff
@@ -294,6 +294,31 @@ pub use sp_std;
 /// for the case when that would create a circular dependency. You usually _do not_ want to add
 /// this flag, as tracing doesn't cost you anything by default anyways (it is added as a no-op)
 /// but is super useful for debugging later.
+///
+/// # Feature patch
+///
+/// Using `feature_force_version`, one can keep using an old host function version by default
+/// in its runtime.
+/// This is usually only for patching runtime that requires additional steps before switching
+/// to a new host function.
+/// Syntax is:
+///
+/// ```
+/// #[sp_runtime_interface::runtime_interface(feature_force_version=old_state,a_function,1)]
+/// trait RuntimeInterface {
+/// 		fn a_function(&mut self) {
+/// 		}
+///
+/// 		#[version(2)]
+/// 		fn a_function(&mut self) {
+/// 		}
+/// }
+/// ```
+///
+/// In this case, when feature `old_state` (warning '-' are not supported in feature name) is
+/// defined, the host function for `RuntimeInterface::root` will be the version 1 even if later
+/// version exists. This functionality is currently only use for migration state and may be
+/// removed in the future.
```

Comment:
IMO we should not bloat the runtime interface macros with all these things.

You can just add a second method `root_v1` or whatever and do the `#[cfg(bla)]` where the code is called.

Reply:
So please revert these changes.

Reply:
Or just make `root` take the state version directly.

Reply:
I am not sure I understand you.
I cannot put the '#[cfg(bla)]' in externalities, I need my runtime to link with the right host function, the feature is only for the runtime.

> Or just make root take the state version directly.

you mean pass the state version as a parameter of the runtime root function, I did implement it like this at some point and I really think it is worse.
You need to read the version (could still read it from runtime-version), then inject it to state machine from client, also cumulus will also need to inject it.
Or you can read the version from the state, but that is also an implementation we drop.

Reply:
Originally, this runtime interface macro was just intended to be here until 'old_state' feature will not be use anymore, I end up making the doc for the review.

I think if we don't want this (can be useful), I would rather try to put  the feature in front of the `#[runtime_interface]` of storage and child storage even if it creates some redundant code.

Reply:
I mean, create the `fn root(changes, state_version)` function.

Reply:
Yes, that is something I implemented before.
But then as I mentioned you need to read state version from runtime in client code, then pass it  in state-machine.
The storage proof need to attach the state version too (so requires changing its format), so cumulus can pass it when verifying pov.

> I think if we don't want this (can be useful), I would rather try to put the feature in front of the #[runtime_interface] of storage and child storage even if it creates some redundant code.

This does not sound that bad to me (for dropping the macro code).